In [1]:
import numpy as np
import pandas as pd
import io
import statsmodels.formula.api as smf
import scipy.stats as stats

In [5]:
data_string = ''' Estimate Std.Error
(Intercept) 3446.830     81.840
totwrk        -0.169      0.018
age            2.714      1.472
south        102.272     41.925
male          87.108     35.173
smsa         -54.188     33.193
yngkid       -13.051     50.459
marr          31.360     42.263
union         11.866     38.186'''
data_string

' Estimate Std.Error\n(Intercept) 3446.830     81.840\ntotwrk        -0.169      0.018\nage            2.714      1.472\nsouth        102.272     41.925\nmale          87.108     35.173\nsmsa         -54.188     33.193\nyngkid       -13.051     50.459\nmarr          31.360     42.263\nunion         11.866     38.186'

In [6]:
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

Estimate  Std.Error
(Intercept)  3446.830     81.840
totwrk         -0.169      0.018
age             2.714      1.472
south         102.272     41.925
male           87.108     35.173
smsa          -54.188     33.193
yngkid        -13.051     50.459
marr           31.360     42.263
union          11.866     38.186

In [7]:
df['znachimost'] = np.abs(df['Estimate']/df['Std.Error'])
df

Estimate  Std.Error  znachimost
(Intercept)  3446.830     81.840   42.116691
totwrk         -0.169      0.018    9.388889
age             2.714      1.472    1.843750
south         102.272     41.925    2.439404
male           87.108     35.173    2.476559
smsa          -54.188     33.193    1.632513
yngkid        -13.051     50.459    0.258646
marr           31.360     42.263    0.742020
union          11.866     38.186    0.310742

In [8]:
alpha = 0.05
t = stats.t.ppf(1-alpha/2, 707-9)
t.round(2)

1.96

In [9]:
df['result'] = df.apply(lambda x: 'znachimo' if x['znachimost'] > t else 'ne znachimo', axis=1 )
df

Estimate  Std.Error  znachimost       result
(Intercept)  3446.830     81.840   42.116691     znachimo
totwrk         -0.169      0.018    9.388889     znachimo
age             2.714      1.472    1.843750  ne znachimo
south         102.272     41.925    2.439404     znachimo
male           87.108     35.173    2.476559     znachimo
smsa          -54.188     33.193    1.632513  ne znachimo
yngkid        -13.051     50.459    0.258646  ne znachimo
marr           31.360     42.263    0.742020  ne znachimo
union          11.866     38.186    0.310742  ne znachimo

In [10]:
df_lab = pd.read_csv ('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Labour.csv')
df_lab

capital  labour     output       wage
0     2.606563     184   9.250759  43.080307
1     1.323237      91   3.664310  27.780016
2    22.093692     426  28.781516  44.467748
3    10.737851      72   4.124642  39.734710
4     1.161365      46   2.890150  34.650709
..         ...     ...        ...        ...
564   2.625403      20   1.424376  33.477545
565   1.276386      61   2.109048  26.300732
566   1.953869     117   6.241870  41.153979
567   1.318527      46   7.902237  66.720139
568   2.832391     161   6.804085  36.618435

[569 rows x 4 columns]

In [11]:
model = smf.ols(data=df_lab, formula='np.log(output)~np.log(capital)+np.log(labour)').fit()
model.summary(alpha=0.01).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [12]:
model.t_test('np.log(capital)=0.5')

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0             0.2076      0.017    -17.014      0.000       0.174       0.241

In [13]:
t_cr = stats.t.ppf(1-0.01/2, 569-3)
t_cr.round(2)

2.58

## 1.2

In [23]:
df  =  pd.read_csv ( 'https://raw.githubusercontent.com/ryupepa/Econometrica/master/python-notebooks/data-csv/sleep75.csv' ) 
df 

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     ...  spwrk75  totwrk  union  worknrm  workscnd  exper  yngkid  yrsmarr  \
0    ...        0    3438      0     3438         0     14       0       13   
1    ...        0    5020      0     5020         0     11       0        0   
2    ...        1    2815      0     2815         0     21       0        0   
3    ...        1    3786      0     3786         0     12       0       12   
4    ...        1    2580      0     2580         0     44       0       33   
..   ...      ...     ...    ...      ...       ...    ...     ...      ...   
701  ...        1    2026      0     2026         0     27       0       18   
702  ...        0     675      1      465       210     18       0        4   
703  ...        1    1851      0     1851         0     19       0       17   
704  ...        1    1961      1     1481       480     31       0       22   
705  ...        0    2363      0     2363         0      8       1        9   

        hrwage  agesq  
0     7.070004   1024  
1     1.429999    961  
2    20.529997   1936  
3     9.619998    900  
4     2.750000   4096  
..         ...    ...  
701        NaN   2025  
702        NaN   1156  
703        NaN   1369  
704        NaN   2916  
705        NaN    900  

[706 rows x 34 columns]

In [31]:
sleep_eq = smf.ols(formula='sleep~totwrk+age+south+male+smsa+yngkid+marr+union+np.log(hrwage)', data=df).fit()
sleep_eq.params.round(3)
sleep_eq.summary(alpha=0.1).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [33]:
alpha = 0.1
t_cr=stats.t.ppf(1-alpha/2, 532-1-9)
t_cr.round(3)

1.648

In [65]:
data_string = '''name 	coef	std.err	t	P>|t|	[0.05	0.95]
Intercept	3431.8036	96.309	35.633	0.000	3273.107	3590.500
totwrk	-0.1581	0.021	-7.576	0.000	-0.192	-0.124
age	2.4371	1.672	1.458	0.146	-0.318	5.192
south	78.0461	45.287	1.723	0.085	3.423	152.669
male	36.4854	43.350	0.842	0.400	-34.946	107.917
smsa	-34.9646	38.232	-0.915	0.361	-97.962	28.033
yngkid	50.1365	54.790	0.915	0.361	-40.145	140.418
marr	54.0716	46.920	1.152	0.250	-23.242	131.386
union	27.0187	43.044	0.628	0.530	-43.908	97.946
np.log(hrwage)	-2.7268	32.202	-0.085	0.933	-55.789	50.336
 '''

In [66]:
df_string = pd.read_csv(io.StringIO(data_string), sep='\s+')
df_string 

name       coef  std.err       t  P>|t|     [0.05     0.95]
0       Intercept  3431.8036   96.309  35.633  0.000  3273.107  3590.500
1          totwrk    -0.1581    0.021  -7.576  0.000    -0.192    -0.124
2             age     2.4371    1.672   1.458  0.146    -0.318     5.192
3           south    78.0461   45.287   1.723  0.085     3.423   152.669
4            male    36.4854   43.350   0.842  0.400   -34.946   107.917
5            smsa   -34.9646   38.232  -0.915  0.361   -97.962    28.033
6          yngkid    50.1365   54.790   0.915  0.361   -40.145   140.418
7            marr    54.0716   46.920   1.152  0.250   -23.242   131.386
8           union    27.0187   43.044   0.628  0.530   -43.908    97.946
9  np.log(hrwage)    -2.7268   32.202  -0.085  0.933   -55.789    50.336

In [68]:
df_string['result'] = df_string.apply(lambda x: 'значимо' if abs(x['t'])>t_cr else 'не значимо',axis=1)
df_string

name       coef  std.err       t  P>|t|     [0.05     0.95]  \
0       Intercept  3431.8036   96.309  35.633  0.000  3273.107  3590.500   
1          totwrk    -0.1581    0.021  -7.576  0.000    -0.192    -0.124   
2             age     2.4371    1.672   1.458  0.146    -0.318     5.192   
3           south    78.0461   45.287   1.723  0.085     3.423   152.669   
4            male    36.4854   43.350   0.842  0.400   -34.946   107.917   
5            smsa   -34.9646   38.232  -0.915  0.361   -97.962    28.033   
6          yngkid    50.1365   54.790   0.915  0.361   -40.145   140.418   
7            marr    54.0716   46.920   1.152  0.250   -23.242   131.386   
8           union    27.0187   43.044   0.628  0.530   -43.908    97.946   
9  np.log(hrwage)    -2.7268   32.202  -0.085  0.933   -55.789    50.336   

       result  
0     значимо  
1     значимо  
2  не значимо  
3  не значимо  
4  не значимо  
5  не значимо  
6  не значимо  
7  не значимо  
8  не значимо  
9  не значимо

- intercept, totwork - значимы
- t_cr = 1.648

## Задача 1.4

In [53]:
df_wage = pd.read_csv ('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/wage2.csv')
df_wage

wage  hours   IQ  KWW  educ  exper  tenure  age  married  black  south  \
0     769     40   93   35    12     11       2   31        1      0      0   
1     808     50  119   41    18     11      16   37        1      0      0   
2     825     40  108   46    14     11       9   33        1      0      0   
3     650     40   96   32    12     13       7   32        1      0      0   
4     562     40   74   27    11     14       5   34        1      0      0   
..    ...    ...  ...  ...   ...    ...     ...  ...      ...    ...    ...   
930   520     40   79   28    16      6       1   30        1      1      1   
931  1202     40  102   32    13     10       3   31        1      0      1   
932   538     45   77   22    12     12      10   28        1      1      1   
933   873     44  109   25    12     12      12   28        1      0      1   
934  1000     40  107   32    12     17      18   35        1      0      1   

     urban  sibs  brthord  meduc  feduc     lwage  
0        1     1      2.0    8.0    8.0  6.645091  
1        1     1      NaN   14.0   14.0  6.694562  
2        1     1      2.0   14.0   14.0  6.715384  
3        1     4      3.0   12.0   12.0  6.476973  
4        1    10      6.0    6.0   11.0  6.331502  
..     ...   ...      ...    ...    ...       ...  
930      0     0      1.0   11.0    NaN  6.253829  
931      1     7      7.0    8.0    6.0  7.091742  
932      0     9      NaN    7.0    NaN  6.287858  
933      0     1      1.0    NaN   11.0  6.771935  
934      0     5      1.0    NaN    NaN  6.907755  

[935 rows x 17 columns]

In [55]:
wage_eq = smf.ols(formula='np.log(wage)~age+IQ+married+south+urban', data=df_wage).fit()
wage_eq.params.round(3)
wage_eq.summary(alpha=0.01).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [56]:
alpha = 0.01
t_cr=stats.t.ppf(1-alpha/2, 935-1-5)
t_cr.round(3)

2.581

In [60]:
data_string = ''' Name coef std.err t  P>|t| [0.005 0.995]
Intercept 4.9740 0.165 30.064 0.000 4.547 5.401
age	0.0213	0.004	5.303	0.000	0.011	0.032
IQ	0.0082	0.001	9.751	0.000	0.006	0.010
married	0.2010	0.040	4.993	0.000	0.097	0.305
south	-0.0990	0.027	-3.690	0.000	-0.168	-0.030
urban	0.1750	0.028	6.331	0.000	0.104	0.246'''   

In [61]:
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df 

Name    coef  std.err       t  P>|t|  [0.005  0.995]
0  Intercept  4.9740    0.165  30.064    0.0   4.547   5.401
1        age  0.0213    0.004   5.303    0.0   0.011   0.032
2         IQ  0.0082    0.001   9.751    0.0   0.006   0.010
3    married  0.2010    0.040   4.993    0.0   0.097   0.305
4      south -0.0990    0.027  -3.690    0.0  -0.168  -0.030
5      urban  0.1750    0.028   6.331    0.0   0.104   0.246

In [69]:
df['result'] = df.apply(lambda x: 'значимо' if abs(x['t'])>t_cr else 'не значимо',axis=1)
df

Name    coef  std.err       t  P>|t|  [0.005  0.995]   result
0  Intercept  4.9740    0.165  30.064    0.0   4.547   5.401  значимо
1        age  0.0213    0.004   5.303    0.0   0.011   0.032  значимо
2         IQ  0.0082    0.001   9.751    0.0   0.006   0.010  значимо
3    married  0.2010    0.040   4.993    0.0   0.097   0.305  значимо
4      south -0.0990    0.027  -3.690    0.0  -0.168  -0.030  значимо
5      urban  0.1750    0.028   6.331    0.0   0.104   0.246  значимо

-  все коэффициенты значимы 
-  t_cr = 2.581

## Задача 1.7

In [73]:
data_string = '''       Estimate  Std.Error
(Intercept)   -7.472      1.063
log(q)         0.838      0.009
log(pl)        0.044      0.106
log(pk)        0.188      0.087
log(pf)        0.713      0.064'''

In [83]:
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

Estimate  Std.Error
(Intercept)    -7.472      1.063
log(q)          0.838      0.009
log(pl)         0.044      0.106
log(pk)         0.188      0.087
log(pf)         0.713      0.064

In [84]:
df['znachimost'] = np.abs(df['Estimate']/df['Std.Error'])
df

Estimate  Std.Error  znachimost
(Intercept)    -7.472      1.063    7.029163
log(q)          0.838      0.009   93.111111
log(pl)         0.044      0.106    0.415094
log(pk)         0.188      0.087    2.160920
log(pf)         0.713      0.064   11.140625

## Задача 1.8

In [85]:
df_e = pd.read_csv ('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Electricity.csv')
df_e

cost        q        pl      sl      pk      sk      pf      sf
0     0.2130      8.0   6869.47  0.3291  64.945  0.4197  18.000  0.2512
1     3.0427    869.0   8372.96  0.1030  68.227  0.2913  21.067  0.6057
2     9.4059   1412.0   7960.90  0.0891  40.692  0.1567  41.530  0.7542
3     0.7606     65.0   8971.89  0.2802  41.243  0.1282  28.539  0.5916
4     2.2587    295.0   8218.40  0.1772  71.940  0.1623  39.200  0.6606
..       ...      ...       ...     ...     ...     ...     ...     ...
153   6.8293    946.6  10642.16  0.0883  43.600  0.1914  51.463  0.7203
154   3.7605    377.0   7432.24  0.2117  74.120  0.2274  33.436  0.5609
155   3.9822    391.0   5826.04  0.1926  78.288  0.0924  44.633  0.7151
156  30.1880   5317.0   9586.63  0.0845  78.008  0.2009  41.840  0.7147
157  67.8562  15220.0   6986.09  0.1451  74.025  0.2596  25.662  0.5945

[158 rows x 8 columns]

In [86]:
electr = smf.ols(formula='np.log(cost)~np.log(q)+np.log(pl)+np.log(pk)+np.log(pf)', data=df_e).fit()
electr.params.round(3)
electr.summary(alpha=0.01).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [88]:
alpha = 0.01
t_cr=stats.t.ppf(1-alpha/2, 158-1-4)
t_cr.round(3)

2.608

In [89]:
data_string = '''name coef	std.err	t	P>|t|	[0.005	0.995]
Intercept	-7.4722	1.063	-7.028	0.000	-10.246	-4.699
np.log(q)	0.8381	0.009	89.284	0.000	0.814	0.863
np.log(pl)	0.0443	0.106	0.418	0.676	-0.232	0.320
np.log(pk)	0.1884	0.087	2.164	0.032	-0.039	0.416
np.log(pf)	0.7131	0.064	11.081	0.000	0.545	0.881'''

In [90]:
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

name    coef  std.err       t  P>|t|  [0.005  0.995]
0   Intercept -7.4722    1.063  -7.028  0.000 -10.246  -4.699
1   np.log(q)  0.8381    0.009  89.284  0.000   0.814   0.863
2  np.log(pl)  0.0443    0.106   0.418  0.676  -0.232   0.320
3  np.log(pk)  0.1884    0.087   2.164  0.032  -0.039   0.416
4  np.log(pf)  0.7131    0.064  11.081  0.000   0.545   0.881

In [91]:
df['result'] = df.apply(lambda x: 'значимо' if abs(x['t'])>t_cr else 'не значимо',axis=1)
df

name    coef  std.err       t  P>|t|  [0.005  0.995]      result
0   Intercept -7.4722    1.063  -7.028  0.000 -10.246  -4.699     значимо
1   np.log(q)  0.8381    0.009  89.284  0.000   0.814   0.863     значимо
2  np.log(pl)  0.0443    0.106   0.418  0.676  -0.232   0.320  не значимо
3  np.log(pk)  0.1884    0.087   2.164  0.032  -0.039   0.416  не значимо
4  np.log(pf)  0.7131    0.064  11.081  0.000   0.545   0.881     значимо

- intercept, log(q), log(pf) - значимы
- t_cr = 2.608

## Задача 2.1

In [92]:
df = pd.read_csv ('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/Labour.csv')
df

capital  labour     output       wage
0     2.606563     184   9.250759  43.080307
1     1.323237      91   3.664310  27.780016
2    22.093692     426  28.781516  44.467748
3    10.737851      72   4.124642  39.734710
4     1.161365      46   2.890150  34.650709
..         ...     ...        ...        ...
564   2.625403      20   1.424376  33.477545
565   1.276386      61   2.109048  26.300732
566   1.953869     117   6.241870  41.153979
567   1.318527      46   7.902237  66.720139
568   2.832391     161   6.804085  36.618435

[569 rows x 4 columns]

In [94]:
model = smf.ols(data=df, formula='np.log(output)~np.log(capital)+np.log(labour)').fit()
model.summary(alpha=0.01).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [95]:
t_nabl = (0.2076-0.5)/0.017
t_cr = stats.t.ppf(1-0.01/2,569-2-1)
print('Коэффициент b_capital = 0.05') if abs(t_nabl)<t_cr else print('Коэффициент b_capital значительно отличается от значения 0.5')

Коэффициент b_capital значительно отличается от значения 0.5


In [96]:
t_cr = stats.t.ppf(1-0.01/2, 569-1-2)
t_cr.round(2)

2.58

In [97]:
t_nabl = (0.2076-0.5)/0.017
t_nabl

-17.2

- t_nabl = -17.2
- t_cr = 2.58
- Коэффициент b_capital значительно отличается от значения 0.5

## Задача 2.2

In [100]:
model = smf.ols(data=df, formula='np.log(output)~np.log(capital)+np.log(labour)').fit()
model.summary(alpha=0.1).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [103]:
t_nabl = (0.7148-0.7)/0.023
t_cr = stats.t.ppf(1-0.1/2,569-2-1)
print('Коэффициент b_labour = 0.7') if abs(t_nabl)<t_cr else print('Коэффициент b_labour значительно отличаетcя от значения 0.7')

Коэффициент b_labour = 0.7


In [106]:
t_nabl = (0.7148-0.7)/0.023
round(t_nabl, 2)

0.64

In [107]:
t_cr = stats.t.ppf(1-0.1/2,569-2-1)
round(t_cr, 2)

1.65

- t_cr = 1.65
- t_nabl = 0.64
- Коэффициент b_labour = 0.7

## Задача 2.5

In [108]:
df = pd.read_csv ('https://raw.githubusercontent.com/ryupepa/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
df

age  black  case  clerical  construc  educ  earns74  gdhlth  inlf  leis1  \
0     32      0     1  0.000000  0.000000    12        0       0     1   3529   
1     31      0     2  0.000000  0.000000    14     9500       1     1   2140   
2     44      0     3  0.000000  0.000000    17    42500       1     1   4595   
3     30      0     4  0.000000  0.000000    12    42500       1     1   3211   
4     64      0     5  0.000000  0.000000    14     2500       1     1   4052   
..   ...    ...   ...       ...       ...   ...      ...     ...   ...    ...   
701   45      0   702  0.182331  0.030075    12     5500       1     0   5069   
702   34      0   703  0.182331  0.030075    10     2500       0     0   5885   
703   37      0   704  0.182331  0.030075    12     3500       1     0   4719   
704   54      0   705  0.182331  0.030075    17    32500       1     0   5149   
705   30      0   706  0.182331  0.030075    16     6750       1     0   4747   

     ...  spwrk75  totwrk  union  worknrm  workscnd  exper  yngkid  yrsmarr  \
0    ...        0    3438      0     3438         0     14       0       13   
1    ...        0    5020      0     5020         0     11       0        0   
2    ...        1    2815      0     2815         0     21       0        0   
3    ...        1    3786      0     3786         0     12       0       12   
4    ...        1    2580      0     2580         0     44       0       33   
..   ...      ...     ...    ...      ...       ...    ...     ...      ...   
701  ...        1    2026      0     2026         0     27       0       18   
702  ...        0     675      1      465       210     18       0        4   
703  ...        1    1851      0     1851         0     19       0       17   
704  ...        1    1961      1     1481       480     31       0       22   
705  ...        0    2363      0     2363         0      8       1        9   

        hrwage  agesq  
0     7.070004   1024  
1     1.429999    961  
2    20.529997   1936  
3     9.619998    900  
4     2.750000   4096  
..         ...    ...  
701        NaN   2025  
702        NaN   1156  
703        NaN   1369  
704        NaN   2916  
705        NaN    900  

[706 rows x 34 columns]

In [109]:
model = smf.ols(data=df, formula='sleep~totwrk+age+south+male+smsa+yngkid+marr+union').fit()
model.summary(alpha=0.01).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [110]:
t_nabl = (87.1081-90)/35.173
t_cr = stats.t.ppf(1-0.01/2,706-8-1)
print('Коэффициент b_male = 90') if abs(t_nabl)<t_cr else print('Коэффициент b_male значительно отличается от значения 90')

Коэффициент b_male = 90


In [111]:
t_nabl = (87.1081-90)/35.173
round(t_nabl, 2)

-0.08

In [112]:
t_cr = stats.t.ppf(1-0.01/2,706-8-1)
round(t_cr ,2)

2.58

- t_cr = 2.58
- t_nabl = -0.08
- Коэффициент b_male = 90